src/core/lifetime.d

```diff
@@ -2361,3 +2368,66 @@ template _d_delstructImpl(T)
     assert(innerDtors == 2);
     assert(outerDtors == 1);
 }
+
+/**
+ * Allocate an exception of type `T` from the exception pool.
+ * It has the same interface as `rt.lifetime._d_newclass()`.
+ * The class type must be Throwable or derived from it,
+ * and cannot be a COM or C++ class. The compiler must enforce this.
+ * Returns:
+ *      default initialized instance of the type
+ */
+Throwable _d_newThrowable(T)() @trusted
+    if (is(T : Throwable))
+{
+    auto ci = T.classinfo;
+
+    debug(PRINTF) printf("_d_newThrowable(ci = %p, %s)\n", ci, cast(char *) T.stringof);
+
+    assert(!(ci.m_flags & TypeInfo_Class.ClassFlags.isCOMclass));
+    assert(!(ci.m_flags & TypeInfo_Class.ClassFlags.isCPPclass));
```

Comment:
I would suggest that the entire reliance on TypeInfo can be gotten rid off because you can do these checks on the compiler side.

Reply:
I'm not sure of this. For example, `ci.initializer.length == 76`, while `T.sizeof == 8` and the code seg faults when I use `T.init` instead of `ci.initializer`. I haven't been able to find where the difference comes from yet.

Reply:
I guess the initializer is the actual size of the object in memory, whereas `T.sizeof` is just the size of the reference.

Reply:
Moreover, when running `dmd/test/runnable/test19317.d`, `T.init == nil`. However, the template is most likely instantiated correctly, because the test passes when the hook uses `TypeInfo_Class`.